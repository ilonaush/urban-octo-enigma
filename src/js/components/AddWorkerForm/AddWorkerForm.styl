.addWorkerForm
  display flex
  flex-direction column
