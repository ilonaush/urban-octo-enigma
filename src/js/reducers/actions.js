export default {
    employWorker: function (worker) {
        return {
            type: 'EMPLOY_WORKER',
            worker: worker
        }
    },
    editWorkTime: function (worker) {
        return {
            type: 'EDIT_WORK_TIME',
            worker: worker,
        }
    },
    fireWorker: function (workerId) {
        return {
            type: 'FIRE_WORKER',
            id: workerId
        }
    }
};
