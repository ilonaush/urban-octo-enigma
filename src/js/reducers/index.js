import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducer   from './reducers';
import  { saveStore, loadStore } from '../services/LocalStorageService';

const predefinedStore = loadStore() || {};

const store = createStore(
    reducer,
    predefinedStore

);

store.subscribe(() => {
    saveStore(store)
});

export {
    store, predefinedStore
};
