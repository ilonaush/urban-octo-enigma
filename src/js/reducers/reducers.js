import {ACTIONS_TYPES} from "../constants";

const initialState = {
  loading: false,
};

let reducer = (state = initialState, action) => {
  switch (action.type) {
    case ACTIONS_TYPES.GET_CATS:
      return {
        ...state,
        cats: action.payload.cats,
        loading: action.payload.loading,
        loadingType: action.payload.loadingType
      };
    case ACTIONS_TYPES.ADD_CAT:
      return {
        ...state,
        cats: action.payload.cats,
        loading: action.payload.loading,
        loadingType: action.payload.loadingType
      };
    case ACTIONS_TYPES.FEED_CAT:
      return {
        ...state,
        cats: action.payload.cats,
        loading: action.payload.loading,
        loadingType: action.payload.loadingType
      };
    case ACTIONS_TYPES.HUG_CAT:
      return {
        ...state,
        cats: action.payload.cats,
        loading: action.payload.loading,
        loadingType: action.payload.loadingType
      };
    case ACTIONS_TYPES.WASH_CAT:
      return {
        ...state,
        cats: action.payload.cats,
        loading: action.payload.loading,
        loadingType: action.payload.loadingType
      };
    case ACTIONS_TYPES.ISSUE_CAT:
      return {
        ...state,
        cats: action.payload.cats,
        loading: action.payload.loading,
        loadingType: action.payload.loadingType
      };
    case ACTIONS_TYPES.SET_LOADING: {
      return {
        ...state,
        loading: action.payload.loading,
        loadingType: action.payload.loadingType
      }
    }
    default:
      return state;
  }
};

export default reducer;
