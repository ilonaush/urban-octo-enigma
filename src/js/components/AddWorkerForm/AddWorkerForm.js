import React, {Component} from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import actions from '../../reducers/actions';
import { bindActionCreators } from "redux";
import {Query} from "../../services/RequestService";
import "./AddWorkerForm.styl";
import {withRouter} from "react-router-dom";

export class AddWorkerForm extends Component {

    constructor(props) {
        super(props);
        this.state = {
            name: '',
            surname: '',
            position: '',
        };

        this.handleChange = this.handleChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.handleLoad = this.handleLoad.bind(this);

    }

    onSubmit(e) {
        e.preventDefault();
        const worker = {
            fullname: `${this.state.name} ${this.state.surname}`,
            position: this.state.position,
            id: Date.now()
        };

        this.props.actions.employWorker(worker);
        const response = this.handleLoad();


    }

    async handleLoad() {
        this.props.handleLoading(true);
        const response = await Query();
        if (response) {
            this.props.handleLoading(false);
            this.props.history.push('/');
        }
    }

    handleChange(value, name) {
        this.setState({
            [name]: value
        })
    }

    render() {
        return (
            <div>
                <form className='addWorkerForm'>
                    <input name='name'
                           placeholder='Name'
                           onChange={(event) => this.handleChange(event.target.value, event.target.name)}
                           value={this.state.name}/>
                    <input name='surname'
                           placeholder='Surname'
                           onChange={(event) => this.handleChange(event.target.value, event.target.name)}
                           value={this.state.surname} />
                    <input name='position'
                           placeholder='Position'
                           onChange={(event) => this.handleChange(event.target.value, event.target.name)}
                           value={this.state.position}/>
                    <button type='submit' onClick={this.onSubmit}>Employ a worker</button>
                </form>
            </div>
        );
    }
}

AddWorkerForm.propTypes = {};


function mapStateToProps(state) {
    return { workers: state.workers  }
}
function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators(actions, dispatch)
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(AddWorkerForm);
