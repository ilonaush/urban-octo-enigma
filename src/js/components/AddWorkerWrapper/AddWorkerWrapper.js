import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import actions from "../../reducers/actions";
import {connect} from "react-redux";
import AddWorkerForm from "../AddWorkerForm/AddWorkerForm";
import  "./AddWorkerWrapper.styl";

export class AddWorkerWrapper extends Component {

    /**
     * handles submit on add WorkerForm, dispatches an action
     * @param e
     * @param cat
     */
    handleSubmit = (e, cat) => {
        e.preventDefault();
        cat = {
            name: `${cat.name}`,
            id: Date.now(),
            age: cat.age,
            color: cat.color,
            reason: cat.reason,
            location: cat.location,
            health: cat.health
        };
        this.props.addCat(cat).then(() => this.props.history.push('/'))
    };

    render() {
        return (
            <Fragment>
                <AddWorkerForm onSubmit={this.handleSubmit}/>
                <img  className='cat-bg' src="/images/cat-bg-2.png" alt=""/>
            </Fragment>
        );
    }
}

AddWorkerWrapper.propTypes = {};

export default connect(
    (state) => ({cats: state.cats}),
    (dispatch) => ({addCat: (cat) => dispatch(actions.addCat(cat))})
)(AddWorkerWrapper);
