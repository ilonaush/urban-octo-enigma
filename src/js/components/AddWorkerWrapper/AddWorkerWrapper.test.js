import React from 'react';
import {AddWorkerWrapper}  from './AddWorkerWrapper';
import {Provider} from "react-redux";
import {store} from "../../reducers/index";
import {mount, shallow} from "enzyme/build";
import renderer from "react-test-renderer";
import AddWorkerForm from "../AddWorkerForm/AddWorkerForm";



describe('AddWorkerWrapper', () => {
    it('renders correctly', () => {
        const AddWorkerWrapperComponent = renderer.create(<AddWorkerWrapper/>);
        expect((AddWorkerWrapperComponent).toJSON()).toMatchSnapshot();
    });

    it('calls submit function  when  the form is submitted', async () => {
        const mockAddCatAction = jest.fn().mockResolvedValue(43);
        const props = {
            addCat: mockAddCatAction
        };
        const AddWorkerWrapperComponent = mount(
            <Provider store={store}>
                <AddWorkerWrapper {...props}/>
            </Provider>
        );

        const AddWorkerFormComponent = AddWorkerWrapperComponent.find(AddWorkerForm);
        expect(mockAddCatAction.mock.calls.length).toBe(0);
        AddWorkerFormComponent.simulate('submit');
        expect(mockAddCatAction.mock.calls.length).toBe(1)
    });


});

