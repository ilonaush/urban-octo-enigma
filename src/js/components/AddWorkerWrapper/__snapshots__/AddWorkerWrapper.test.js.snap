// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddWorkerWrapper renders correctly 1`] = `
Array [
  <form
    className="addWorkerForm"
    onSubmit={[Function]}
  >
    <h3>
      Please share some info about the cat with us
    </h3>
    <input
      id="name-input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      value=""
    />
    <input
      name="age"
      onChange={[Function]}
      placeholder="Age"
      value=""
    />
    <input
      name="color"
      onChange={[Function]}
      placeholder="Color"
      value=""
    />
    <textarea
      name="reason"
      onChange={[Function]}
      placeholder="Reason"
      value=""
    />
    <input
      name="location"
      onChange={[Function]}
      placeholder="Location"
      value=""
    />
    <input
      name="health"
      onChange={[Function]}
      placeholder="Describe health"
      value=""
    />
    <button
      id="submit-btn"
      type="submit"
    >
      Accept the cat
    </button>
  </form>,
  <img
    alt=""
    className="cat-bg"
    src="/images/cat-bg-2.png"
  />,
]
`;
